### PR TITLE
Create update job timeout processor

### LINKED ISSUE #15036: Create `UpdateJobTimeoutProcessor`
- [x] Create the `UpdateJobTimeoutProcessor`
    - Goes into the `io.camunda.zeebe.engine.processing.job` package
    - Register the processor in the `JobEventProcessors`
    - Needs to have access to the (immutable) `JobState`
- [x] Implement the command handling in the processor
    - Find a job in the state with the provided job key
        - If it can't find any write a `NOT_FOUND` response and rejection
    - Find the job deadline in the state using the Job we just got from the state
        - If it can't find any write a `INVALID_STATE` response and rejection
    - Write the `JobIntent.TIMEOUT_UPDATED` event
- [x] Unit tests

### PR DESCRIPTION
**Pull request by ana-vinogradova-camunda**
## Description

Creating `JobUpdateTimeoutProcessor` to calculate new deadline.

If there is no job found for the job key then a NOT_FOUND response and rejection are written
If there is no deadline for the job with the specified job key then INVALID_STATE response and rejection are written

## Related issues

closes #15036 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutProcessor.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutProcessor.java:29]: ❌ So I (finally) had a think about the error message (and a chat with Nico 😄). We propose to change it to:

`Expected to update the timeout of job with key '%d', but it is not active`

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java:29]: ❌ I thought of another test case I'd like to see, which is updating the timeout multiple times:

1. create a job
2. Activate the job
3. Update the timeout
4. Update the timeout again
5. Assert that the deadline is the current time + the timeout of step 4

I believe this test is relevant because this way we verify that we are actually updating the deadline in the Job ColumnFamily, as well as the deadlines ColumnFamily. If we wouldn't the second update timeout would be rejected.

Comment (remcowesterhoud) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java:29]: I want 2 more test cases in the `JobTimeOutTest`. This test class verifies we are timing out jobs correctly.

**`shouldTimeOutAfterDecreasingTimeout`**

1. Create a job
2. Activate job with a high timeout
3. Update job to a lower timeout
4. Increase the time
5. Verify the job has expired

**`shouldTimeOutAfterIncreasingTimeout`**

1. Create a job
2. Activate job with a high timeout
3. Update job to an even higher timeout
4. Increase the time
5. Verify the job has expired

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks @ana-vinogradova-camunda 🙇 I have played a comment with the definitive error message I'd like to see.

I also thought of some more test cases I'd like to see to verify timeouts still behave as expected. Please have a look and let me know if you have any questions!

File: engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutProcessor.java
Comment (ana-vinogradova-camunda) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutProcessor.java]: Done 👍

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java
Comment (ana-vinogradova-camunda) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java:31]: Done 👍

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java
Comment (koevskinikola) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java:113]: :+1: Nice test coverage!

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java
Comment (koevskinikola) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java:31]: :x: We should also add some tenant-related tests:
* Ensure that a job belonging to a "custom" tenant, can have it's deadline updated (with an authorization of the same tenant).
* Ensure that a job belonging to a "custom-A" tenant, **can't** have it's deadline updated by a "custom-B" tenant.

Similar tests are implemented [here](https://github.com/camunda/zeebe/blob/ff15090290739f5932aa17c145a1bad9d0664841/engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateRetriesTest.java#L163-L207).

**Review by koevskinikola — DISMISSED** [review]
Nice work @ana-vinogradova-camunda! :rocket: 

I have two additional suggestions and it should be ready for merging.

Once you have the Java client code merged, you can also add IT tests in the `MultiTenancyOverIdentityIT` test. You can find some [similar test cases in the same class](https://github.com/camunda/zeebe/blob/d6eaafadfb411f1b30a87022658691ec82c4630c/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityIT.java#L759-L842).

File: engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutProcessor.java
Comment (ana-vinogradova-camunda) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutProcessor.java]: Thank you, I've updated the method to be [JobState#getJob(key, authorizations)](https://github.com/camunda/zeebe/blob/062daa45cbd5dba6e96e95d8cdcaee89ae30a83c/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/JobState.java#L34)

Comment (tmetzke) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutProcessor.java]: Looks good now 👍

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java
Comment (tmetzke) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java]: 🔧 To improve readability and consistency, the new test methods could move above the helper method.

This is not blocking, it's just a matter of style. I'd suggest to improve this while we're at it.

**Review by tmetzke — APPROVED** [review]
Awesome, @ana-vinogradova-camunda 👍 

The changes since the last review look good, I'd say this is close to ready.
I have one last little nitpick, I'm already approving the PR to unblock you here.

**Comment by tmetzke** [comment]
> Once you have the Java client code merged, you can also add IT tests in the `MultiTenancyOverIdentityIT` test. You can find some [similar test cases in the same class](https://github.com/camunda/zeebe/blob/d6eaafadfb411f1b30a87022658691ec82c4630c/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityIT.java#L759-L842).

@ana-vinogradova-camunda and I decided to move this into a follow-up issue rather than further blocking this PR here to wait for the Java client PR to be merged.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java
Comment (ana-vinogradova-camunda) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java]: Absolutely agree, thank you.
This file is marked as `Outdated` and I've checked that the new test methods are above the helper method.

Comment (tmetzke) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobUpdateTimeoutTest.java]: Funny, the last commit I can see now wasn't there when I reviewed this. Strange, well, maybe the page was cached. It looks good, we can resolve this here 👍

**Comment by ana-vinogradova-camunda** [comment]
bors merge

**Review by korthout — APPROVED** [review]
How about this?

bors merge

**Comment by korthout** [comment]
I see no option to [dismiss](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/dismissing-a-pull-request-review) remco's review

Screenshot 2023-11-21 at 11 09 04

@ana-vinogradova-camunda I think you'll have to recreate the pull request 😞

**Comment by ana-vinogradova-camunda** [comment]
Closing this PR as the merged is blocked.

**Pull request closed without merging**
